### DISCUSSION TITLE
Window is not defined! > bug

### DISCUSSION THREAD

**Discussion by ucimo** [react-static/react-static#1600] [Q&A]
I was amaized how quickly I could move all of my pure React code into React-Static and it worked as expected. _Great stuff._ 

**But** then I tried building it and I keep [bumping into this](https://stackoverflow.com/questions/66944586/react-static-build-error-err-require-esm-or-referenceerror-window-is-not-defin). 

How would you say that I escape this node.js execution of browser code? I tried adding return statements if window or document is undefined, but without any results. I read everything I could find on react-static, but I'm probably lacking knowledge of node.js to fix this properly.

Please let me know if some specific info is needed. Here's my [package.json](https://prnt.sc/119uq0u)

**Comment by SleeplessByte** [comment]
This doesn't really have much to do with node.js or esm. What's happening is that during a production build, each page is built using "server-side rendering". In that context, window, for example, doesn't exist. 

All code and packages that rely on browser specific APIs (for example web APIs) should be conditionally excluded. Most implementation do something like:

```js
if (typeof window !== 'undefined') {
  // ... require, use code that needs window to exist
} 
```

Note that this is not specific to react-static but is a limitation on all SSR solutions unless window is mocked-in (as a no-op polyfill).

**Reply by ucimo** [reply]
My terminology and wording is probably bad, so I get why you're writing me this, but I know window doesn't exist on the server side (node.js). What I'm saying is that I've already added this line `if(typeof window === "undefined") return null;` before I use window or document anywhere in my project and I still get this error. It means I use some npm package that relies on window object or browser API and that's causing the problem? If so, how can I know where to condition this package? I already built this app months ago and now I want to adapt it to 'react-static' to get some SSO. If I were building it from start in 'react-stat' I would have a lot better idea where to look for this issue, as I would have continuously made builds after bigger changes in the app. Since that's not the case, I have to deal with the complexity of a finished app, so I was hoping someone could give me a push in the right direction.

**Reply by SleeplessByte** [reply]
> It means I use some npm package that relies on window object or browser API and that's causing the problem?

Yep that can definitely be the case.

> If so, how can I know where to condition this package? 

In the stackoverflow issue you only show the final error message, but the stacktrace likely has all the information including the package that's causing the error. Once you find it, you can do various things, one is dynamically importing (using `import(...).then`) the script or using `require`. Using a regular `import {} from` won't work because it's _hoisted_ to the top of the file.

Another method would be to use `patch-package` to make the code of the source package check for `window`, but I only use that as a last resort.

**Reply by ucimo** [reply]
I can't solve this. If you could give me any direction where to look or what to do, some blog-post, some tutorial, anything. I'm completely lost here. Here's another look at my errors: [error1](https://prnt.sc/125uaoa), [error2](https://prnt.sc/125ucu7), [error3](https://prnt.sc/125udgf). Is there anyone else who could help me or any other place where comuntiy around this framework could explain to me how to read this?

**Reply by SleeplessByte** [reply]
It says you're using [`RAFManager`](https://www.npmjs.com/package/raf-manager) which literally tries to access `window` in the first line of code:

```javascript
function () {
	let lastTime = 0;
	let vendors = ['ms', 'moz', 'webkit', 'o'];
	for (let x = 0; x < vendors.length && !window.requestAnimationFrame; ++x) {
		window.requestAnimationFrame = window[vendors[x] + 'RequestAnimationFrame'];
		window.cancelAnimationFrame = window[vendors[x] + 'CancelAnimationFrame']
			|| window[vendors[x] + 'CancelRequestAnimationFrame'];
	}
}());
```

The mention of the cause of this error is also in your first screenshot and second screenshot. You can't just drop libraries in and expect them to work isomorphically (both on the browser and the server).

My recommendation here would be to either conditionally load the RAF Manager, or not use it at all. It seems an outdated thing,  and it's very unlikely that you need the polyfill, nor do you need the library itself and can manager your animation frames directly using `requestAnimationFrame` and `clearAnimationFrame`.
